### PR TITLE
Vector encoder

### PR DESCRIPTION
**Pull request by KWillets**
SSE vector-based encoder.  Encodes a quad of uint's at a time.

This is for non-Delta only.  The vector encoder is added to streamvbyte.c, and shuffle tables are moved to an include due to bulk.

### CODE REVIEW

File: perf.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include "streamvbyte.h"
+static inline void rdtsc(unsigned long long * destination)
+{
+  uint64_t t;
+  __asm__ volatile (".byte 0x0f, 0x31" : "=A" (t)) ;
+  *destination = t;
+}
+
+int main() {
+	int N = 500000;
```
Comment (lemire) [inline perf.c]: Ok. For compression it makes sense to go RAM-to-RAM.

File: src/encode_poc.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
+
+#include <x86intrin.h>
+#include <stdio.h>
+#include "shuffle_tables.h"
+
+// POC standalone test until code looks ok
+// usage:  gcc -msse4.1 -o encode_poc encode_poc.c
+
+void dump(__m128i x, char *msg) {
```
Comment (lemire) [inline src/encode_poc.c]: Testing code should be moved to something like the ``tests`` subdirectory. I.e., we don't want the regular users to get this in their binaries.

File: src/encode_poc.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		uint8_t *__restrict__ keyPtr, uint8_t *__restrict__ dataPtr,
+		uint32_t count) {
+
+  uint8_t * outData = dataPtr;
+  uint8_t * outKey  = keyPtr;
+
+  for (uint32_t c = 0; c <= count - 4; c += 4)
+    outData += streamvbyte_encode4((u128 *) (in + c), outData, outKey++);
+
+  // todo:  integrate and make this valid:
```
Comment (lemire) [inline src/encode_poc.c]: Not sure what this todo stands for...?

File: src/encode_poc.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  //  outData = svb_encode_scalar(in + c, outKey, outData, count - c);
+
+  return outData;
+}
+
+// ugly test code
```
Comment (lemire) [inline src/encode_poc.c]: This should be moved to the "test" package. The "ugliness" of the testing routines is not critical, but we don't want them to be part of the main codebase.

File: src/streamvbyte.c [hunk 5]
```diff
@@ -343,13 +73,82 @@ static uint8_t *svb_encode_scalar(const uint32_t *in,
...
 
+#ifdef __AVX__
+
+typedef union M128 {
+  char i8[16];
+  uint32_t u32[4];
+  __m128i i128;
+} u128;
+
+const u128 Ones = {.i8 = {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1}};
```
Comment (lemire) [inline src/streamvbyte.c]: A global constant is unwarranted. This is literally used only in one function. It is much safer to declare it within the function where it is used.

File: Makefile [hunk 6]
```diff
@@ -5,9 +5,9 @@
 
 PROCESSOR:=$(shell uname -m)
 ifeq ($(PROCESSOR), aarch64) 
-CFLAGS = -fPIC -std=c99 -O3 -Wall -Wextra -pedantic -Wshadow
+CFLAGS = -msse4.1 -fPIC -std=c99 -O3 -Wall -Wextra -pedantic -Wshadow
```
[the remark this answers is not in the document]
Comment (KWillets) [reply Makefile]: Oops, I did figure out that it was using the other clause.  It seems fine without the flag.

File: src/encode_poc.c (hunk 3 above)
Comment (KWillets) [reply src/encode_poc.c]: done already.  This file will be removed.

**Comment by lemire** [comment]
Ok. Merging. I might do additional minor changes after testing on various systems.

**Comment by lemire** [comment]
I added one fix for your code. Doing ``*(__m128i*)`` is undefined behavior that will actually result in a crash on some hardware.

**Comment by KWillets** [comment]
I did hit a few issues with alignment during load/stores.  Probably we should not assume input is aligned.

**Comment by lemire** [comment]
The current version should be a lot safer. I do know of any undefined behavior remaining.

**Comment by lemire** [comment]
(I do not know of any undefined behavior remaining.)

**Comment by marshallpierce** [comment]
This was pretty interesting to figure out while I was porting to the Rust version. I'm new to SIMD; what would be good resources to learn more about how to come up with these clever chains of SIMD instructions? I've found Daniel Kusswurm's book "Modern X86 Assembly Language Programming", and of course there's Agner Fog's site. Anything else I should look into?

**Comment by lemire** [comment]
I don't know about Kusswurm's book. I have co-written a bunch of papers that you might find interesting...

* https://arxiv.org/abs/1709.07821
* https://arxiv.org/abs/1611.07612
* https://arxiv.org/abs/1401.6399

**Comment by KWillets** [comment]
Those papers are lodged somewhere in my mind.  It's useful to see that a given instruction is good for a given domain; for instance small table lookups via pshufb are invaluable.  

Another method is to look for single-word bit hacks that will work in SIMD lanes -- the SIMD instructions don't seem to add much below the word or byte level, so older methods have to be fit into the basic SIMD instructions like compare, shift and multiply.

I also made a lot of use of this reference:  https://software.intel.com/sites/landingpage/IntrinsicsGuide/ .

It's very handy for narrowing down instructions by type and generation.

**Comment by lemire** [comment]
One thing I'd add is that focusing specifically on Intel intrinsics is maybe not the best all-around strategy. There are nice things being done at ARM. Sadly, it suffers from much poorer documentation,

**Comment by KWillets** [comment]
I love ARM; I highly recommend the ARM SVE paper.  

NEON looks like a good next project.

**Comment by lemire** [comment]
See https://github.com/lemire/streamvbyte/issues/6

**Pull request merged**